src/attribution/onCLS.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getLoadState} from '../lib/getLoadState.js';
+import {getSelector} from '../lib/getSelector.js';
+import {onCLS as unattributedOnCLS} from '../onCLS.js';
+import {CLSAttribution, CLSReportCallback, CLSReportCallbackWithAttribution, CLSMetric, CLSMetricWithAttribution, ReportOpts} from '../types.js';
+
+
+const getLargestLayoutShiftEntry = (entries: LayoutShift[]) => {
+  return entries.reduce((a, b) => a && a.value > b.value ? a : b);
+}
+
+const getLargestLayoutShiftSource = (sources: LayoutShiftAttribution[]) => {
+  return sources.find((s) => s.node && s.node.nodeType === 1) || sources[0];
+}
+
+const attributeCLS = (metric: CLSMetric): CLSMetricWithAttribution => {
+  let attribution: CLSAttribution = {};
+  if (metric.entries.length) {
+    const largestEntry = getLargestLayoutShiftEntry(metric.entries);
+    if (largestEntry && largestEntry.sources && largestEntry.sources.length) {
+      const largestSource = getLargestLayoutShiftSource(largestEntry.sources);
+      if (largestSource) {
+        attribution = {
+          largestShiftTarget: getSelector(largestSource.node),
+          largestShiftTime: largestEntry.startTime,
+          largestShiftValue: largestEntry.value,
+          largestShiftSource: largestSource,
+          largestShiftEntry: largestEntry,
```

Comment:
Just checking: this Entry would already be in the entries array, so this is just a convenience?

I'm just curious if the other attributions should expose "the most important entry" even if its an array size 1.  FID and FCP that seems a bit pointless, but for LCP/INP it seems reasonable?

Reply:
I think it could be convenient.

Also, I can see a world where FID and LCP return more than a single Entry in the `entries` list, and so this would still have value in revealing the "main" one (probably always the first in list):

- For FID, it could be all the events of the interaction.
- For LCP, it could be all previous candidates.